sensing/pointcloud_preprocessor/src/outlier_filter/ring_outlier_filter_nodelet.cpp

```diff
@@ -44,82 +44,116 @@ RingOutlierFilterComponent::RingOutlierFilterComponent(const rclcpp::NodeOptions
+    const uint16_t ring = *reinterpret_cast<const uint16_t *>(&input->data[data_idx + ring_offset]);
+    ring2indices[ring].push_back(data_idx);
+  }
+
+  // walk range: [walk_first_idx, walk_last_idx]
+  int walk_first_idx = 0;
+  int walk_last_idx = -1;
+
+  for (const auto & indices : ring2indices) {
+    if (indices.size() < 2) continue;
+
+    walk_first_idx = 0;
+    walk_last_idx = -1;
+
+    for (size_t idx = 0U; idx < indices.size() - 1; ++idx) {
+      const size_t & current_data_idx = indices[idx];
+      const size_t & next_data_idx = indices[idx + 1];
+      walk_last_idx = idx;
 
       // if(std::abs(iter->distance - (iter+1)->distance) <= std::sqrt(iter->distance) * 0.08)
-      const auto current_pt_azimuth =
-        *reinterpret_cast<float *>(&input_ptr->data[current_idx + azimuth_offset]);
-      const auto next_pt_azimuth =
-        *reinterpret_cast<float *>(&input_ptr->data[next_idx + azimuth_offset]);
-      float azimuth_diff = next_pt_azimuth - current_pt_azimuth;
+
+      const float & current_azimuth =
+        *reinterpret_cast<const float *>(&input->data[current_data_idx + azimuth_offset]);
+      const float & next_azimuth =
+        *reinterpret_cast<const float *>(&input->data[next_data_idx + azimuth_offset]);
+      float azimuth_diff = next_azimuth - current_azimuth;
       azimuth_diff = azimuth_diff < 0.f ? azimuth_diff + 36000.f : azimuth_diff;
 
-      const auto current_pt_distance =
-        *reinterpret_cast<float *>(&input_ptr->data[current_idx + distance_offset]);
-      const auto next_pt_distance =
-        *reinterpret_cast<float *>(&input_ptr->data[next_idx + distance_offset]);
+      const float & current_distance =
+        *reinterpret_cast<const float *>(&input->data[current_data_idx + distance_offset]);
+      const float & next_distance =
+        *reinterpret_cast<const float *>(&input->data[next_data_idx + distance_offset]);
 
       if (
-        std::max(current_pt_distance, next_pt_distance) <
-          std::min(current_pt_distance, next_pt_distance) * distance_ratio_ &&
+        std::max(current_distance, next_distance) <
+          std::min(current_distance, next_distance) * distance_ratio_ &&
         azimuth_diff < 100.f) {
-        continue;
+        continue;  // Determined to be included in the same walk
       }
-      if (isCluster(input_ptr, tmp_indices)) {
-        for (const auto & tmp_idx : tmp_indices) {
-          output_modifier.push_back(
-            std::move(*reinterpret_cast<PointXYZI *>(&input_ptr->data[tmp_idx])));
+
+      if (isCluster(
+            input, indices[walk_first_idx], indices[walk_last_idx],
+            walk_last_idx - walk_first_idx + 1)) {
+        for (int i = walk_first_idx; i <= walk_last_idx; i++) {
+          auto output_ptr = reinterpret_cast<PointXYZI *>(&output.data[output_size]);
+          *output_ptr = *reinterpret_cast<const PointXYZI *>(&input->data[indices[i]]);
+          output_size += output.point_step;
```

Comment:
This part assume data layout is `PointXYZI` or at least that input memory layout starts the same than `PointXYZI`. In practice it will just copy the first 4 fields. Is there a reason to ignore the other fields from input (ring, azimut, etc?) I know today Autoware use the ring filter as the last filter that needs "extended" fields, but who knows if it will still be true in the future?

Reply:
At the moment, it is fine with no changes (cf. https://github.com/autowarefoundation/autoware.universe/blob/c7d84c64aadb6dc86a95920f392c5036225d76a0/common/autoware_point_types/include/autoware_point_types/types.hpp#L59-L63)

But we have a bug and are dealing with it in https://github.com/autowarefoundation/autoware.universe/pull/2618.

In the long-term perspective, we will make chagne according to [this design](https://github.com/autowarefoundation/autoware-documentation/blob/main/docs/design/autoware-architecture/sensing/data-types/point-cloud.md).

Reply:
I see. Thank you for the link. I find the manual 4-bytes padding quite strange... but that's another topic.
Still, I think this way of using `reinterpret_cast` is UB. For example, since the the compiler is not told not to pad the 2 structs with annotations such as `__attribute__((packed))`, the compiler is "free" to pad the structs as it sees fit.

Reply:
It is a universal problem in Autoware and not related to this Pull Request, so please raise another Issue.